src/pad.rs

```diff
@@ -0,0 +1,363 @@
+            self.digestinfo_prefix.len() + self.digest_alg.output_len;
+
+        // Require at least 8 bytes of padding. Since we disallow keys smaller
+        // than 2048 bits, this should never happen anyway.
+        debug_assert!(out.len() >= digest_len + 11);
+        let pad_len = out.len() - digest_len - 3;
+        out[0] = 0;
+        out[1] = 1;
+        for i in 0..pad_len {
+            out[2 + i] = 0xff;
+        }
+        out[2 + pad_len] = 0;
+
+        let (digest_prefix, digest_dst) = out[3 + pad_len..].split_at_mut(
+            self.digestinfo_prefix.len());
+        digest_prefix.copy_from_slice(self.digestinfo_prefix);
+        digest_dst.copy_from_slice(
+            digest::digest(self.digest_alg, msg).as_ref());
+        Ok(())
+    }
+
+    fn verify(&self, msg: untrusted::Input, encoded: untrusted::Input)
+              -> Result<(), ()> {
+        encoded.read_all((), |decoded| {
+            if try!(decoded.read_byte()) != 0 ||
+               try!(decoded.read_byte()) != 1 {
+                return Err(());
+            }
+
+            let mut ps_len = 0;
+            loop {
+                match try!(decoded.read_byte()) {
+                    0xff => { ps_len += 1; },
+                    0x00 => { break; },
+                    _ => { return Err(()); }
+                }
+            }
+            if ps_len < 8 {
+                return Err(());
+            }
+
+            let decoded_digestinfo_prefix =
+                try!(decoded.skip_and_get_input(
+                        self.digestinfo_prefix.len()));
+            if decoded_digestinfo_prefix != self.digestinfo_prefix {
+                return Err(());
+            }
+
+            let digest_alg = self.digest_alg;
+            let decoded_digest =
+                try!(decoded.skip_and_get_input(digest_alg.output_len));
+            let digest = digest::digest(digest_alg, msg.as_slice_less_safe());
+            if decoded_digest != digest.as_ref() {
+                return Err(());
+            }
+            Ok(())
+        })
+    }
+
+
+}
+
+// Maximum supported length of the salt;
+const MAX_SLEN: usize = 512 / 8;
+
+// Maxmimum supported length for intermediate value used in PSS padding.
+const MAX_MLEN: usize = 512;
+
+pub struct PSSPadding {
+    pub digest_alg: &'static digest::Algorithm,
+    pub mgf: fn(seed: &[u8], mask_len: usize, mask: &mut [u8],
+                digest_alg: &'static digest::Algorithm) -> Result<(), ()>,
+    pub s_len: usize
+}
+
+impl PaddingAlgorithm for PSSPadding {
+    // Implement padding procedure per EMSA-PSS,
+    // https://tools.ietf.org/html/rfc3447#section-9.1.
+    fn encode(&self, msg: &[u8], out: &mut [u8], rng: &rand::SecureRandom)
+              -> Result<(), ()> {
+
+        //    Steps:
+
+        //    1.  If the length of M is greater than the input limitation for the
+        //        hash function (2^61 - 1 octets for SHA-1), output "message too
+        //        long" and stop.
```

Comment:
Regarding these comments that quote the spec: At a minimum, let's get rid of the excess indention at the start and rewrap the text to 80 chars. However, I actually would prefer them to be abbreviated to just "Steps 1-3". The reader can read the actual steps in the spec. This way, we don't have to ensure that the text of the actual quotes are correct. See `ring::ec::suite_b::ecdh::ecdh` for an example of the style I mean.